packages/story-editor/src/app/rightClickMenu/useLayerSelect.js

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import {
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+} from '@web-stories-wp/react';
+import SAT from 'sat';
+import { __ } from '@web-stories-wp/i18n';
+import { Icons } from '@web-stories-wp/design-system';
+import { trackEvent } from '@web-stories-wp/tracking';
+
+/**
+ * Internal dependencies
+ */
+import useStory from '../story/useStory';
+import createPolygon from '../canvas/utils/createPolygon';
+import { useCanvas } from '../canvas';
+import { getDefinitionForType } from '../../elements';
+
+function useLayerSelect({ menuItemProps, menuPosition, isMenuOpen }) {
+  const [isSubMenuOpen, setIsSubMenuOpen] = useState(false);
+  const { nodesById } = useCanvas(({ state: { nodesById } }) => ({
+    nodesById,
+  }));
+  const { currentPage, selectedElements, setSelectedElementsById } = useStory(
+    ({
+      state: { currentPage, selectedElements },
+      actions: { setSelectedElementsById },
+    }) => ({
+      currentPage,
+      selectedElements,
+      setSelectedElementsById,
+    })
+  );
+
+  useEffect(() => {
+    // Close submenu if the menu itself also closes.
+    if (!isMenuOpen) {
+      setIsSubMenuOpen(false);
+    }
+  }, [isMenuOpen]);
+
+  const { x, y } = menuPosition;
+  const getIntersectingElements = useCallback(() => {
+    if (!currentPage?.elements?.length || !Object.keys(nodesById)?.length) {
+      return [];
+    }
+    const clickedPoint = new SAT.Vector(x, y);
+    return currentPage.elements
+      .map((element) => {
+        const { id, rotationAngle } = element;
+        const node = nodesById[id];
+        if (!node) {
+          return null;
+        }
+        const { x: elX, y: elY, width, height } = node.getBoundingClientRect();
+        const elementP = createPolygon(rotationAngle, elX, elY, width, height);
+        return SAT.pointInPolygon(clickedPoint, elementP) ? element : null;
+      })
+      .filter((el) => el);
+  }, [currentPage, x, y, nodesById]);
```

Comment:
This is cool. @littlemilkstudio do you think this should use a `ResizeObserver`? May be hard to tell if it's necessary without doing some performance auditing.

May not be necessary for now and we can always follow up later.